tests/test_downloader_handlers.py

```diff
@@ -733,6 +735,16 @@ def test_download_with_proxy_https_timeout(self):
         timeout = yield self.assertFailure(d, error.TimeoutError)
         self.assertIn(domain, timeout.osError)
 
+    def test_download_with_proxy_without_http_scheme(self):
+        def _test(response):
+            self.assertEqual(response.status, 200)
+            self.assertEqual(response.url, request.url)
+            self.assertEqual(response.body, b'http://example.com')
+
+        http_proxy = self.getURL('').replace('http:', '')
```

Comment:
Shouldn’t we test without the `//`? Or at least test both ways?

```suggestion
        http_proxy = self.getURL('').replace('http://', '')
```

Reply:
So I did try this out as well, and the test actually doesn't give the correct result. I looked into it and found that while running the test with `http_proxy = self.getURL('').replace('http://', '')` this [line 19](https://github.com/scrapy/scrapy/blob/master/scrapy/core/downloader/webclient.py#L19) in `webclient` throws an exception indicating that it wasn't able to extract the hostname. It seems like the reason is `urllib.parse.urlparse` implementation

```
>>> from urllib.parse import urlparse
>>> parsed = urlparse('test-uri.com')
>>> parsed.hostname
>>> parsed.hostname == None
True
parsed = urlparse('//test-uri.com')
>>> parsed.hostname
'test-uri.com'
```

Reply:
Then we need to change the implementation to support `test-uri.com`. It would be great to also support `//test-uri.com`, but I would consider that optional, given that users are much more likely to use `test-uri.com`.

Reply:
It actually works fine on the user-end, I created the following spider and ran it. 

```
import scrapy


class ExampleSpider(scrapy.Spider):
    name = 'example'

    def start_requests(self):
        proxies = ["example-proxy.com:9999", "//example-proxy.com:9999", "http://example-proxy.com:9999"]
        for proxy in proxies:
            yield scrapy.Request(url='http://quotes.toscrape.com/page/1/', callback=self.parse, meta={"proxy": proxy})

    def parse(self, response):
        print('Finished Successfully')
```

The `httpproxy` middleware is able to parse the proxy url correctly because of [these 2 lines](https://github.com/scrapy/scrapy/blob/master/scrapy/downloadermiddlewares/httpproxy.py#L32-L33) (_urlunparse()_ seems to adds the '//'). 

Not sure if the request made in the unit test passes through this middleware, but I think the fix should be as simple as checking if the url contains a scheme and if not adding a '//' to the front in `webclient._parse`. What do you think?

Reply:
> I think the fix should be as simple as checking if the url contains a scheme and if not adding a '//' to the front in webclient._parse. What do you think?

Sounds good to me.

Reply:
Looks like it's not as easy as it sounds, tried a couple of fixes:- 

1. checking `urlparse(url).scheme` and prepending  '//' if missing
2. checking if `urlparse(url).netloc` was identified and prepending '//' if missing
 
and was unsuccessful. From the [urllib.parse.urlparse](https://docs.python.org/3/library/urllib.parse.html?highlight=urlsplit#urllib.parse.urlparse) docs:-

> Following the syntax specifications in [RFC 1808](https://tools.ietf.org/html/rfc1808.html), urlparse recognizes a netloc only if it is properly introduced by ‘//’. Otherwise the input is presumed to be a relative URL and thus to start with a path component.

`urlparse` thus produces a few different results for various types of URIs (without schema) making it hard to introduce a simple 'fix'

```
>>> urlparse('127.0.0.1:11586')
ParseResult(scheme='', netloc='', path='127.0.0.1:11586', params='', query='', fragment='')
>>> urlparse('127.0.0.1:11586/')
ParseResult(scheme='127.0.0.1', netloc='', path='11586/', params='', query='', fragment='')
>>> urlparse('://127.0.0.1:11586/')
ParseResult(scheme='', netloc='', path='://127.0.0.1:11586/', params='', query='', fragment='')
>>> urlparse('://127.0.0.1:11586')
ParseResult(scheme='', netloc='', path='://127.0.0.1:11586', params='', query='', fragment='')
>>> urlparse('//127.0.0.1:11586')
ParseResult(scheme='', netloc='127.0.0.1:11586', path='', params='', query='', fragment='')
>>> urlparse('//127.0.0.1:11586/')
ParseResult(scheme='', netloc='127.0.0.1:11586', path='/', params='', query='', fragment='')
>>> urlparse('example-proxy.com')
ParseResult(scheme='', netloc='', path='example-proxy.com', params='', query='', fragment='')
>>> urlparse('example-proxy.com:3333')
ParseResult(scheme='', netloc='', path='example-proxy.com:3333', params='', query='', fragment='')
>>> urlparse('example-proxy.com:3333/')
ParseResult(scheme='example-proxy.com', netloc='', path='3333/', params='', query='', fragment='')
>>> urlparse('://example-proxy.com:3333/')
ParseResult(scheme='', netloc='', path='://example-proxy.com:3333/', params='', query='', fragment='')
>>> urlparse('://example-proxy.com:3333')
ParseResult(scheme='', netloc='', path='://example-proxy.com:3333', params='', query='', fragment='')
>>> urlparse('//example-proxy.com:3333')
ParseResult(scheme='', netloc='example-proxy.com:3333', path='', params='', query='', fragment='')
```

I personally think it would be easier for the developer to enforce the standard quoted above ([RFC 1808](https://tools.ietf.org/html/rfc1808.html)) while writing the test, otherwise (if necessary) we could open a new issue to work on it.

Reply:
The original issue is that users enter the domain name without a protocol, so we cannot expect them to add `//`.

What about prepending `//` if the URL does not match `^\w+://`?

Reply:
This issue doesn't seem to exist when a user creates a spider and runs it, only seems to exist while writing unit tests

> It actually works fine on the user-end, I created the following spider and ran it.
> 
> ```
> import scrapy
> 
> 
> class ExampleSpider(scrapy.Spider):
>     name = 'example'
> 
>     def start_requests(self):
>         proxies = ["example-proxy.com:9999", "//example-proxy.com:9999", "http://example-proxy.com:9999"]
>         for proxy in proxies:
>             yield scrapy.Request(url='http://quotes.toscrape.com/page/1/', callback=self.parse, meta={"proxy": proxy})
> 
>     def parse(self, response):
>         print('Finished Successfully')
> ```
> 
> The `httpproxy` middleware is able to parse the proxy url correctly in [these 2 lines](https://github.com/scrapy/scrapy/blob/master/scrapy/downloadermiddlewares/httpproxy.py#L32-L33) (found while debugging that `urlunparse` prepends the '//' while unparsing the `_parse_proxy` result).

Will try out your suggestion, Thanks

Reply:
I’m guessing the reason it happens in the test is because it happens with `localhost` as proxy. You could argue that `localhost` being the proxy is uncommon, but still I think we should support it, specially if it’s trivial to support.